Unset session `started_at` in Application Server

johanstokking opened the pull request:
<!--
Thanks for submitting a pull request. Please fill the template below,
otherwise we will not be able to process this pull request.
-->

#### Summary
<!--
A short summary, referencing related issues:
Closes #0000, References #0000, etc.
-->

Unset session `started_at`

Addresses #4766

#### Changes
<!-- What are the changes made in this pull request? -->

- Do not set session `started_at` in AS
- Unset `started_at` as part of `as-db cleanup`

#### Testing

<!-- How did you verify that this change works? -->

1. Made local device
2. Temporarily allowed setting `session.started_at` in AS (by allowing field mask validation, not part of this PR)
3. Set the started at to a value: 
```
TTN_LW_NETWORK_SERVER_ENABLED=false ttn-lw-cli dev set app1 dev1 --session.started_at "2021-01-01T12:00:00Z" --session.dev_addr ABCDABCD
```
4. Decode the value from the database:
```json
{"ids":{"device_id":"dev1","application_ids":{"application_id":"app1"},"dev_eui":"1122334455667788","join_eui":"0000000000000000","dev_addr":"ABCDABCD"},"created_at":"2021-10-27T08:34:35.761063Z","updated_at":"2021-10-27T08:37:21.740897Z","session":{"dev_addr":"ABCDABCD","keys":{},"started_at":"2021-01-01T12:00:00Z"}}
```
5. Run `as-db cleanup`
6. Decode the value from the database:
```json
{"ids":{"device_id":"dev1","application_ids":{"application_id":"app1"},"dev_eui":"1122334455667788","join_eui":"0000000000000000","dev_addr":"ABCDABCD"},"created_at":"2021-10-27T08:34:35.761063Z","updated_at":"2021-10-27T08:44:47.863279Z","session":{"dev_addr":"ABCDABCD","keys":{},"started_at":"0001-01-01T00:00:00Z"}}
```
Notice that it's reset to an empty value, the best we can do now

##### Regressions

<!-- Please indicate features that this change could affect and how that was tested. -->

None, unless clients were relying on faulty behavior

#### Notes for Reviewers
<!--
NOTE: This section is optional.

Motivate briefly why it is implemented this way, if that deviates from the
implementation proposal in the referenced issues.
- How should your reviewers approach this pull request?
- @mention reviewers with special requests or questions for them
-->

@kschiffer I'm leaving the target issue open and will provide some additional information on what the Console should do when merging sessions.

#### Checklist
<!-- Make sure that this pull request is complete. -->

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

johanstokking committed:
as: Do not set session started_at

pgalic96 reviewed (commented):
The changes LGTM. However, I assume we will need to run this only once and then there will be no non-zero `started_at` values anymore in the registry (might be a wrong assumption). Then it shouldn't be necessary to run this periodically with the rest of the cleanup, and maybe we can hide it behind a flag (`--unset-dev-fields` or something like that).

adriansmares commented:
> The changes LGTM. However, I assume we will need to run this only once and then there will be no non-zero `started_at` values anymore in the registry (might be a wrong assumption). Then it shouldn't be necessary to run this periodically with the rest of the cleanup, and maybe we can hide it behind a flag (`--unset-dev-fields` or something like that).

Or as part of a migration, albeit I do agree that we need to this would be longer if we do that.

johanstokking commented:
> The changes LGTM. However, I assume we will need to run this only once and then there will be no non-zero `started_at` values anymore in the registry (might be a wrong assumption). Then it shouldn't be necessary to run this periodically with the rest of the cleanup, and maybe we can hide it behind a flag (`--unset-dev-fields` or something like that).

Best would be to have some sort of migration version of the AS end device registry.

@adriansmares don't we have something like this for NS?

adriansmares commented:
> @adriansmares don't we have something like this for NS?

We do:
https://github.com/TheThingsNetwork/lorawan-stack/blob/8edf8c81c5e8e771e778b67fb8acf4696ee0205f/cmd/ttn-lw-stack/commands/ns_db.go#L468-L491

What needs to happen for that to be useful:
- Bring `as-db migrate` from TTI, make the migration there have version 1
- Mark this cleanup as version 2
- If the recorded version is lower than `asredis.SchemaVersion`, run the missing migrations in order

htdvisser reviewed (commented):
Nothing to add here. I agree that this belongs in a migration, and not in the periodic cleanup command.

johanstokking committed:
cli: Extract Redis schema versioning

johanstokking committed:
cli: Return error when purge fails

johanstokking committed:
cli,as: Migrate database and record version

johanstokking commented:
@adriansmares I replaced the cleanup logic by the `as-db migrate` command that looks a lot like NS's. I updated the original comment and verified that the migration produces the desired outcome.

Output:

```
$ go run ./cmd/ttn-lw-stack as-db migrate
INFO	Connecting to Redis database...
INFO	Setting schema version	{"version": 1}
DEBUG	1 keys migrated
DEBUG	No new version available
```

And again:

```
$ go run ./cmd/ttn-lw-stack as-db migrate
INFO	Connecting to Redis database...
INFO	Existing database schema version	{"version": 1}
INFO	Database schema version is already in latest version
DEBUG	No new version available
```

adriansmares commented on cmd/ttn-lw-stack/commands/as_db.go:
```diff
@@ -29,6 +33,76 @@ var (
+				logger := logger.WithField("key", k)
+				switch {
+				case uidRegexp.MatchString(k):
+					if err := cl.Watch(ctx, func(tx *redis.Tx) error {
```
I'd do a `LockedWatch` here, otherwise we race with the real Application Server instances.

johanstokking committed:
cli: Use locked watch

johanstokking committed:
cli: Run Application Server database migration in all namespaces

johanstokking committed:
cli: Declare device UID regex explicitly
